### PR TITLE
zstd: free Decoder resources when Reset is called with a nil io.Reader

### PR DESCRIPTION
**Pull request by mostynb**
Attempt at fixing #296.

### CODE REVIEW

**Comment by mostynb** [comment]
(I will dig into the test failures in the next few days.)

File: zstd/decoder.go [hunk 1]
```diff
@@ -156,8 +156,14 @@ func (d *Decoder) Reset(r io.Reader) error {
 	if d.current.err == ErrDecoderClosed {
 		return d.current.err
 	}
+
+	d.drainOutput()
+
 	if r == nil {
-		return errors.New("nil Reader sent as input")
+		d.current.b = []byte{}
```
Comment (klauspost) [inline zstd/decoder.go:163]: This should already be handled by the `drainOutput`, so this should be an un-needed alloc.
```suggestion
```

File: zstd/decoder.go [hunk 2]
```diff
@@ -156,8 +156,14 @@ func (d *Decoder) Reset(r io.Reader) error {
...
+		d.current.err = io.EOF
```
Comment (klauspost) [inline zstd/decoder.go:164]: This would make a read seem like the payload was a valid 0 byte block. 

We want to make it clear that a read was performed with a nil input.

```Go
	// ErrNilInput is returned when a nil Reader was given
	// and a non-Reset operation was attempted. 
	ErrNilInput = errors.New("nil input provided as reader")
```

```suggestion
		d.current.err = ErrNilInput
```

**Review by klauspost — COMMENTED** [review]
Thanks for the contribution.

Let's update the docs as well.
```Go
// Reset will reset the decoder the supplied stream after the current has finished processing.
// Note that this functionality cannot be used after Close has been called.
// Reset can be called with a nil reader to release references to the previous reader.
// When sending a nil reader no other operations than Reset or DecodeAll or Close should be used.
func (d *Decoder) Reset(r io.Reader) error {
```

A test that also tests the basic functionality, ie create a Reader, Reset with nil and check that reads returns the expected error and finally that it can be used with a proper reader would be great.

File: zstd/decoder.go (hunk 2 above)
Comment (mostynb) [reply zstd/decoder.go:164]: I guess we should also return this sentinel error if the Decoder was created with a nil reader.

File: zstd/decoder.go (hunk 1 above)
Comment (mostynb) [reply zstd/decoder.go]: Removed in the fixup.

File: zstd/decoder.go (hunk 2 above)
Comment (mostynb) [reply zstd/decoder.go]: Added this change, but with the name ErrDecoderNilInput in the fixup. Maybe that's too specific though?

Comment (klauspost) [reply zstd/decoder.go]: It's fine. Thanks!

**Comment by klauspost** [comment]
Thanks. Looks good. I will give it a final look through and merge when I have the time.
